### DISCUSSION TITLE
Installing Cardinal

### DISCUSSION THREAD

**Discussion by amerallaf** [idaholab/moose#27342] [Q&A Getting Started] [answered]
Hello,

I have been trying to install ardinal with MOOSE's conda Environment, I am folllowing the instructions in the below link:
https://cardinal.cels.anl.gov/with_conda.html

However, I am ending up having the following outcome:

ardinal is using HDF5 from    /home/amer/miniforge
Cardinal is using MOOSE from   /home/amer/cardinal/contrib/moose
Cardinal is using OpenMC from  /home/amer/cardinal/contrib/openmc
Cardinal is using DAGMC from   /home/amer/cardinal/contrib/DAGMC
Cardinal is using Moab from    /home/amer/cardinal/contrib/moab

Cardinal is compiled with the following MOOSE modules
  FLUID_PROPERTIES
  HEAT_TRANSFER
  NAVIER_STOKES
  REACTOR
  SOLID_PROPERTIES
  STOCHASTIC_TOOLS
  THERMAL_HYDRAULICS

Linking libpng: -L/home/amer/miniforge/envs/moose/lib -lpng16
Rebuilding symlinks in /home/amer/cardinal/contrib/moose/framework/build/header_symlinks
Creating Unity Directory /home/amer/cardinal/contrib/moose/framework/build/unity_src
Copying default MOOSE configuration to: /home/amer/cardinal/contrib/moose/framework/include/base/MooseConfig.h...
cp: relocation error: /lib/x86_64-linux-gnu/libacl.so.1: symbol getxattr version ATTR_1.0 not defined in file libattr.so.1 with link time reference
Compiling C++ (in opt mode) /home/amer/cardinal/contrib/moose/framework/contrib/pcre/src/pcrecpp.cc...
make: *** [/home/amer/cardinal/contrib/moose/framework/moose.mk:158: /home/amer/cardinal/contrib/moose/framework/include/base/MooseConfig.h] Error 127
make: *** Waiting for unfinished jobs....
mv: relocation error: /lib/x86_64-linux-gnu/libacl.so.1: symbol getxattr version ATTR_1.0 not defined in file libattr.so.1 with link time reference

What should I do to make it work?

**Comment by milljm** [comment]
The error does seem to be related to altering `LD_LIBRARY_PATH`. It may be that the system you are on contains a wildly different `libattr.so` version.

Can you try the following, see if setting `LD_LIBRARY_PATH` is not necessary?

Exit your terminal (to remove what we asked you to do with the exporting of this variable), and then do:
```bash
cd cardinal
export ENABLE_NEK=false
export HDF5_ROOT=$CONDA_PREFIX
make -j8
```
basically everything except exporting LD_LIBRARY_PATH...

**Reply by amerallaf** [reply]
Oh thank you for your help, this seems to allowed me completing the installation. 

But there is no cardinal-opt. I typed:
./cardinal-opt --help
and I am getting
-bash: ./cardinal-opt: No such file or directory

What else do you suggest me to do?

**Reply by milljm** [reply]
If `./cardinal-opt` isn't there, then I am not sure the build actually completed successfully. Can you copy/paste the output on the screen when you run the following:
```bash
conda activate moose    # if not already activated
cd ~/cardinal
make
```

**Reply by amerallaf** [reply]
Please check the attached two text files, the first output "output1" I got (before performing "make" again), and the second after doing what you asked me.
[output1.txt](https://github.com/idaholab/moose/files/14937147/output1.txt)
[output2.txt](https://github.com/idaholab/moose/files/14937149/output2.txt)

**Reply by milljm** [reply]
Yeah that's better... I actually see the same error in output1.txt. It just happened early on, yet the build process continued for a ways.

This now looks like some indifference with what Conda moose-wasp package you have installed, and what your `cardinal/contrib/moose` requires. It's possible you're using too new a version of our Conda packages, while Cardinal's MOOSE submodule is out of date.

Can you do the following and report back the results:
```bash
conda activate moose    # again, if not already activated :)
cd ~/cardinal/contrib/moose/scripts
./versioner.py wasp --yaml | grep install
conda list | grep moose
```

**Reply by amerallaf** [reply]
Thank you so much for your help, this what I got:
after 
./versioner.py wasp --yaml | grep install
install: moose-wasp=2024.03.15=build_0

after 
conda list | grep moose
moose-dev                 2024.01.23              build_0    https://conda.software.inl.gov/public
moose-libmesh             2024.01.23              build_0    https://conda.software.inl.gov/public
moose-libmesh-vtk         9.2.6                   build_8    https://conda.software.inl.gov/public
moose-mpich               4.0.2                  build_15    https://conda.software.inl.gov/public
moose-peacock             2023.11.29           h1bca318_0    https://conda.software.inl.gov/public
moose-petsc               3.20.3                  build_0    https://conda.software.inl.gov/public
moose-tools               2023.12.20           h55675f2_0    https://conda.software.inl.gov/public
moose-wasp                2023.11.29              build_0    https://conda.software.inl.gov/public

**Reply by milljm** [reply]
Excellent. That was our issue. Just a mismatch with what version is required. In order to fix this, this should do the trick:

```bash
conda activate moose
conda install moose-dev=2024.04.03
```
If this command fails, please stop, and let us know before proceeding!

The issue with doing this however, is that you will need to perform a repository clean, and attempt a new build:
```bash
cd ~/cardinal
git submodule foreach --recursive git clean -xfd
git clean -xfd


# With a clean repo, and updated Conda packages, attempt to build
export ENABLE_NEK=false
export HDF5_ROOT=$CONDA_PREFIX
make -j 8
```

**Reply by amerallaf** [reply]
I did as you instructed and it solved the previous error(s), however, new errors showed up, therefore, no cardinal-opt was created.
Please check the attached text, the errors I could find are located in the end of the text file.

 [output3.txt](https://github.com/idaholab/moose/files/14938052/output3.txt)

**Reply by milljm** [reply]
Looks like the error is:
```pre
/home/amer/cardinal/install/include/openmc/position.h:8:10: fatal error: fmt/format.h: No such file or directory
    8 | #include "fmt/format.h"
      |          ^~~~~~~~~~~~~~
Compiling C++ (in opt mode) /home/amer/cardinal/build/unity_src/mesh_Unity.C...
compilation terminated.
make: *** [/home/amer/cardinal/contrib/moose/framework/build.mk:151: /home/amer/cardinal/build/unity_src/auxkernels_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
make: *** Waiting for unfinished jobs....
In file included from /home/amer/cardinal/install/include/openmc/particle_data.h:6,
                 from /home/amer/cardinal/install/include/openmc/particle.h:12,
                 from /home/amer/cardinal/install/include/openmc/bank.h:6,
                 from /home/amer/cardinal/build/header_symlinks/OpenMCProblemBase.h:30,
                 from /home/amer/cardinal/build/header_symlinks/OpenMCCellAverageProblem.h:21,
                 from /home/amer/cardinal/src/base/OpenMCCellAverageProblem.C:21,
                 from /home/amer/cardinal/build/unity_src/base_Unity.C:11:
/home/amer/cardinal/install/include/openmc/position.h:8:10: fatal error: fmt/format.h: No such file or directory
    8 | #include "fmt/format.h"
      |          ^~~~~~~~~~~~~~
compilation terminated.
make: *** [/home/amer/cardinal/contrib/moose/framework/build.mk:150: /home/amer/cardinal/build/unity_src/base_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
```
@aprilnovak  you wouldn't happen to know about this `fmt/format.h` file? Is this a reason to require LD_LIBRARY_PATH? I would expect for a header file that we would instead be needing to set C_INCLUDE_PATH or some such.

**Reply by milljm** [reply]
Oh wait, this is interesting:

```pre
-- Found fmt: /usr/local/lib/cmake/fmt (version 8.0.1)
```
Just going off of naming schemes here. And guessing that `fmt` is related to `fmt/format.h`. So this may indicate that we may need to set LD_LIBRARAY_PATH to point to Conda's libs.

And now that am seeing that in the log, I am seeing several other cases where things in your `/usr/local/` area are being used instead of Conda's.

Can you run the following tool and report back those results?
```bash
cd ~/cardinal/contrib/moose/scripts
./diagnostics.sh
```

**Reply by amerallaf** [reply]
Thank you again, the attched text file shows the outcome of what you asked me to try.

[output4.txt](https://github.com/idaholab/moose/files/14948134/output4.txt)

**Reply by milljm** [reply]
Thats one of the more clean outputs of diagnotics I've seen. I am not detecting anything out of the ordinary.

I suspect you have third party tools installed in `/usr/local` that are getting in the way however. I am not sure why there would be an issue with `fmt`. From what I am finding on the internet, its a pretty simple library. We can try to install it, then attempt to build Cardinal again. But I suspect we will run into another error of the same nature (some other binary in /usr/local next getting in our way). We will deal with what ever next comes up, in the same manner!

Regardless, doesn't hurt to try. Please perform the following:
```bash
conda activate moose  # if not already
conda install fmt
cd ~/cardinal
git submodule foreach --recursive git clean -xfd
git clean -xfd
export ENABLE_NEK=false
export HDF5_ROOT=$CONDA_PREFIX
make -j 4            # looks like you have 4 available cores according to diagnostic output
```


It is necessary to run the clean commands (and thus start over), because I suspect OpenMC is using `pkg-config` to locate the necessary paths to link fmt properly. Consider the following as an example I ran into on my machine:
```pre
$ pkg-config --libs --cflags fmt
Package fmt was not found in the pkg-config search path.
Perhaps you should add the directory containing `fmt.pc'
to the PKG_CONFIG_PATH environment variable
No package 'fmt' found

$ conda install fmt
$ pkg-config --libs --cflags fmt
-L/data/milljm/host_conda/envs/moose/lib -lfmt
```

**Reply by aprilnovak** [reply]
It was a while ago that we needed to add that LD_LIBRARY_PATH aspect for the OpenMC builds with conda. I _think_ it was related to OpenMC's submodules for fmt, etc. conflicting with system versions (for instance, see this issue: https://github.com/openmc-dev/openmc/issues/2536). Perhaps @pshriwise has a better memory than me?

**Reply by amerallaf** [reply, marked as answer]
Hello @milljm and @aprilnovak, now I have Cardinal working well.

So what I did in the end is uninstalling and reinstalling my Ubuntu (WSL), and start from zero.

I installed MOOSE with Conda environment, based on the following link:
https://mooseframework.inl.gov/getting_started/installation/conda.html

Then I did follow what @milljm instructed me previously:
export ENABLE_NEK=false
export HDF5_ROOT=$CONDA_PREFIX
make -j 8

So **no** "export LD_LIBRARY_PATH=$CONDA_PREFIX/lib:$LD_LIBRARY_PATH"

And now I have a working Cardinal.

Thank you so much @milljm for your time and help.

**Reply by milljm** [reply]
I was going to suggest that moving away from Ubuntu 18 might help, and that WSL makes that an easy process... What Ubuntu version did you end up going with? I am curious so that we may use this information in future discussions.

**Reply by amerallaf** [reply]
The current version is Ubuntu 22.04.3 LTS, previously it was 18.04.6 LTS.
